File: core/src/main/java/com/jetbrains/tmp/learning/SupportedLanguages.java
package com.jetbrains.tmp.learning;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

public enum SupportedLanguages {
    ASM32("asm32", "asm32", "main32.asm", "#", null, null),
    ASM64("asm64", "asm64", "main64.asm", "#", null, null),
    C("C", "c", "main.c", "//", null, null),
    CLOJURE("Clojure", "clojure", "main.clj", ";;", null, null),
    CPP("C++", "c++", "main.cpp", "//", null, null),
    CPP_11("C++11", "c++11", "main_11.cpp", "//", null, null),
    HASKELL("Haskell", "haskell", "Main.hs", "--", null, null),
    HASKELL_7_10("Haskell 7.10", "haskell 7.10", "Main_7.10.hs", "--", null, null),
    HASKELL_8_0("Haskell 8.0", "haskell 8.0", "Main_8.0.hs", "--", null, null),
    JAVA("Java", "java", "Main.java", "//", new String[]{"class Main {"}, new String[]{"}"}),
    JAVA8("Java 8", "java8", "Main.java", "//", new String[]{"class Main {"}, new String[]{"}"}),
    JAVASCRIPT("JavaScript", "javascript", "main.js", "//", null, null),
    MONO_CS("Mono c#", "mono c#", "main.cs", "//", null, null),
    OCTAVE("Octave", "octave", "main.m", "%", null, null),
    PYTHON3("Python 3", "python3", "main.py", "#", null, null),
    R("R", "r", "main.r", "#", null, null),
    RUBY("Ruby", "ruby", "main.rb", "#", null, null),
    RUST("Rust", "rust", "main.rs", "//", null, null),
    SHELL("Shell", "shell", "main.sh", "#", null, null),
    SCALA("Scala", "scala", "Main.scala", "//", null, null),
    INVALID("invalid", "", "", "", null, null);

    private static Map<String, SupportedLanguages> map;
    private final String name;
    private final String comment;
    private final String mainFileName;
    private final String[] beforeCode;
    private final String[] afterCode;
    private final String title;

    /**
     * A Constructor for a supported language.
     *
     * @param name         Language name
     * @param mainFileName A name for main file without extension
     * @param beforeCode   Part of the surrounding code before text
     * @param afterCode    Part of the surrounding code after text
     */
    SupportedLanguages(
            @NotNull String title,
            @NotNull String name,
            @NotNull String mainFileName,
            @NotNull String comment,
            @Nullable String[] beforeCode,
            @Nullable String[] afterCode) {
        this.title = title;
        this.name = name;
        this.mainFileName = mainFileName;
        this.comment = comment;
        this.beforeCode = beforeCode;
        this.afterCode = afterCode;
    }

    @NotNull
    public static SupportedLanguages langOf(@NotNull String lang) {
        if (map == null) {
            map = new HashMap<>();
            Arrays.stream(values()).forEach(value -> map.put(value.getName(), value));
        }

        SupportedLanguages language = map.get(lang);
        if (language == null) {
            return INVALID;
        }

        return language;
    }

    @NotNull
    public String getName() {
        return name;
    }

    @NotNull
    public String getComment() {
        return comment;
    }

    @NotNull
    public String getMainFileName() {
        return mainFileName;
    }

    @NotNull
    public String[] getBeforeCode() {
        return beforeCode != null ? beforeCode : new String[0];
    }

    @NotNull
    public String[] getAfterCode() {
        return afterCode != null ? afterCode : new String[0];
    }

    /**
     * Checks whether the comment string
     *
     * @param line String to check
     * @return true if line is commented, false otherwise
     */
    public boolean isCommentedLine(@NotNull String line) {
        return line.trim().startsWith(comment);
    }

    @NotNull
    @Override
    public String toString() {
        return title;
    }
}

File: stepik-core/src/main/kotlin/org/stepik/core/SupportedLanguages.kt
package org.stepik.core

import org.stepik.core.testFramework.runners.JavaRunner
import org.stepik.core.testFramework.runners.KotlinRunner
import org.stepik.core.testFramework.runners.PythonRunner
import org.stepik.core.testFramework.runners.Runner
import org.stepik.core.testFramework.runners.StubRunner

enum class SupportedLanguages constructor(val title: String,
                                          val langName: String,
                                          val mainFileName: String,
                                          val testFileName: String,
                                          val comment: String,
                                          val beforeCode: String? = null,
                                          val afterCode: String? = null,
                                          private val nextVersion: SupportedLanguages? = null,
                                          val runner: Runner = StubRunner) {
    ASM32("asm32", "asm32", "main32.asm", "test32.asm", "#"),
    ASM64("asm64", "asm64", "main64.asm", "test64.asm", "#"),
    C("C", "c", "main.c", "test.c", "//"),
    CLOJURE("Clojure", "clojure", "main.clj", "test.clj", ";;"),
    CPP("C++", "c++", "main.cpp", "test.cpp", "//"),
    CPP_11("C++11", "c++11", "main_11.cpp", "test_11.cpp", "//"),
    GO("Go", "go", "main.go", "test.go", "//"),
    HASKELL("Haskell 7.8", "haskell", "Main.hs", "Test_7_8.hs", "--",
            "module Main where\nmain = print $ \"Hello, world!\""),
    HASKELL_7_10("Haskell 7.10", "haskell 7.10", "Main.hs", "Test_7_10.hs", "--",
            "module Main where\nmain = print $ \"Hello, world!\""),
    HASKELL_8_0("Haskell 8.0", "haskell 8.0", "Main.hs", "Test_8_0.hs", "--",
            "module Main where\nmain = print $ \"Hello, world!\""),
    JAVA8("Java 8", "java8", "Main.java", "Test.java", "//", "class Main {", "}", runner = JavaRunner()),
    JAVA7("Java 7", "java", "Main.java", "Test.java", "//", "class Main {", "}", JAVA8, runner = JavaRunner()),
    JAVASCRIPT("JavaScript", "javascript", "main.js", "test.js", "//"),
    KOTLIN("Kotlin", "kotlin", "Main.kt", "Test.kt", "//", runner = KotlinRunner()),
    MONO_CS("Mono c#", "mono c#", "main.cs", "test.cs", "//"),
    OCTAVE("Octave", "octave", "main.m", "test.m", "%"),
    PASCAL_ABC("PascalABC.NET", "pascalabc", "Main.pas", "Test.pas", "//"),
    PYTHON3("Python 3", "python3", "main.py", "test.py", "#", runner = PythonRunner()),
    R("R", "r", "main.r", "test.r", "#"),
    RUBY("Ruby", "ruby", "main.rb", "test.rb", "#"),
    RUST("Rust", "rust", "main.rs", "test.rs", "//"),
    SCALA("Scala", "scala", "Main.scala", "test.scala", "//"),
    SHELL("Shell", "shell", "main.sh", "test.sh", "#"),
    INVALID("invalid", "invalid", "", "", "");

    fun isCommentedLine(line: String) = line.trimStart { it.isWhitespace() }.startsWith(comment)

    override fun toString() = title

    fun canUpgradedTo(language: SupportedLanguages): Boolean {
        return language == this || nextVersion?.canUpgradedTo(language) ?: false
    }

    fun comment(string: String) = "$comment$string"

    companion object {

        private val nameMap by lazy {
            values().associateBy { it.langName }
        }

        private val titleMap by lazy {
            values().associateBy { it.title }
        }

        fun langOfName(lang: String) = nameMap.getOrDefault(lang, INVALID)

        fun langOfTitle(lang: String) = titleMap.getOrDefault(lang, INVALID)
    }
}
